### PR TITLE
Feature/router

### PR DESCRIPTION
**Pull request by matipojo**

### CODE REVIEW

File: assets/dev/js/modules/imports/component.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
+import Module from './module';
+
+export default class extends Module {
+	__construct( args ) {
+		if ( ! args || ! args.context ) {
```
Comment (matipojo) [inline assets/dev/js/modules/imports/component.js]: Current manager

File: assets/dev/js/modules/imports/component.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+			throw Error( 'context is required' );
+		}
+
+		this.context = args.context;
+		this.tabs = {};
+		this.isActive = {};
+		this.isModal = false;
```
Comment (matipojo) [inline assets/dev/js/modules/imports/component.js]: Move to ModalComponent

File: assets/dev/js/modules/imports/component.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+	registerRoute( route, callback ) {
+		elementorCommon.route.register( this, route, callback );
+	}
+
+	registerTabRoute( tab ) {
+		this.registerRoute( tab, () => this.activateTab( tab ) );
+	}
+
+	dependency() {
```
Comment (matipojo) [inline assets/dev/js/modules/imports/component.js]: Try remove

File: core/common/modules/finder/assets/js/component.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+		this.isModal = true;
+	}
+
+	getNamespace() {
+		return 'finder';
+	}
+
+	getRoutes() {
+		return {
+			'': () => { /* Nothing to do, it's already rendered. */ },
```
Comment (matipojo) [inline core/common/modules/finder/assets/js/component.js]: move to modal component

File: core/common/modules/finder/assets/js/component.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+				dependency: () => {
+					return this.getItemsView().$activeItem;
+				},
+
+			},
+		};
+	}
+
+	open() {
+		if ( ! this.layout ) {
```
Comment (matipojo) [inline core/common/modules/finder/assets/js/component.js]: rename to modal, move to modal component

File: core/common/modules/finder/assets/js/component.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+			this.layout = new FinderLayout();
+			this.layout.getModal().on( 'hide', () => this.close() );
```
Comment (matipojo) [inline core/common/modules/finder/assets/js/component.js]: test on admin side

File: modules/history/assets/js/history/component.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+			redo: () => this.context.navigate( true ),
+		};
+	}
+
+	getShortcuts() {
+		return {
+			undo: {
+				keys: 'ctrl+z',
+				exclude: [ 'input' ],
+				scopes: [ 'preview', 'panel' ],
```
Comment (matipojo) [inline modules/history/assets/js/history/component.js]: create component preview

File: assets/dev/js/editor/components/template-library/component.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+export default class extends elementorModules.Component {
+	__construct( args ) {
+		super.__construct( args );
+
+		this.isModal = true;
+		this.docLibraryConfig = elementor.config.document.remoteLibrary;
+
+		if ( 'block' === this.docLibraryConfig.type ) {
+			this.setDefault( 'templates/blocks' );
```
Comment (matipojo) [inline assets/dev/js/editor/components/template-library/component.js]: setRouteDefault

File: assets/dev/js/editor/components/template-library/component.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+		} else {
+			this.setDefault( 'templates/pages' );
+		}
+	}
+
+	getNamespace() {
+		return 'library';
+	}
+
+	getTabs() {
```
Comment (matipojo) [inline assets/dev/js/editor/components/template-library/component.js]: initialTabs

File: assets/dev/js/editor/components/template-library/component.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+	}
+
+	activateTab( tab ) {
+		elementorCommon.route.saveState( 'library' );
+
+		super.activateTab( tab );
+	}
+
+	open() {
+		if ( ! this.context.layout ) {
```
Comment (matipojo) [inline assets/dev/js/editor/components/template-library/component.js]: modal

File: assets/dev/js/editor/components/template-library/component.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+		this.context.layout.getModal().hide();
+
+		return true;
+	}
+
+	show( args ) {
+		this.context.modalConfig = args;
+
+		if ( args.toDefault || ! elementorCommon.route.restoreState( 'library' ) ) {
+			elementorCommon.route.to( this.getDefault() );
```
Comment (matipojo) [inline assets/dev/js/editor/components/template-library/component.js]: getRouteDefault

File: assets/dev/js/editor/components/template-library/manager.js [hunk 12]
```diff
@@ -119,10 +88,6 @@ TemplateLibraryManager = function() {
 		return templateTypes;
 	};
 
-	this.getScreens = function() {
-		return screens;
-	};
-
 	this.registerTemplateType = function( type, data ) {
```
Comment (matipojo) [inline assets/dev/js/editor/components/template-library/manager.js]: pro: getScreens - add patch to check if router exist

File: assets/dev/js/editor/components/template-library/manager.js [hunk 13]
```diff
@@ -375,32 +301,43 @@ TemplateLibraryManager = function() {
...
+			},
+		} );
 	};
 
 	this.showTemplates = function() {
-		var activeSource = self.getFilter( 'source' );
+		// The tabs should exist in DOM on loading.
+		self.layout.setHeaderDefaultParts();
+
+		self.loadTemplates( function() {
```
Comment (matipojo) [inline assets/dev/js/editor/components/template-library/manager.js]: es6

File: assets/dev/js/editor/components/template-library/manager.js [hunk 14]
```diff
@@ -375,32 +301,43 @@ TemplateLibraryManager = function() {
...
+			var templatesToShow = self.filterTemplates();
+
+			self.layout.showTemplatesView( new TemplateLibraryCollection( templatesToShow ) );
+		} );
+	};
 
-		var templatesToShow = templatesCollection.filter( function( model ) {
+	this.filterTemplates = function() {
```
Comment (matipojo) [inline assets/dev/js/editor/components/template-library/manager.js]: es6

File: assets/dev/js/editor/regions/panel/pages/editor/component.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+	openEditor( model, view ) {
+		const title = elementor.translate( 'edit_element', [ elementor.getElementData( model ).title ] ),
+		editor = elementor.getPanelView().setPage( 'editor', title, {
+			model: model,
+			controls: elementor.getElementControls( model ),
+			editedElementView: view,
+		} );
+
+		const action = 'panel/open_editor/' + model.get( 'elType' );
```
Comment (matipojo) [inline assets/dev/js/editor/regions/panel/pages/editor/component.js]: // BC + deprecated

File: assets/dev/js/modules/imports/component.js (hunk 3 above)
Comment (matipojo) [reply assets/dev/js/modules/imports/component.js]: It's required

File: core/common/modules/finder/assets/js/component.js (hunk 5 above)
Comment (matipojo) [reply core/common/modules/finder/assets/js/component.js]: moved to modal component

File: assets/dev/js/editor/components/template-library/component.js (hunk 10 above)
Comment (matipojo) [reply assets/dev/js/editor/components/template-library/component.js]: moved to component-modal

File: core/common/assets/js/utils/route.js [hunk 16]
```diff
@@ -56,15 +57,14 @@ export default class extends Commands {
 			return false;
 		}
 
-		const parts = route.split( '/' ),
-			container = parts[ 0 ];
+		const component = this.getComponent( route );
```
Comment (KingYes) [inline core/common/assets/js/utils/route.js]: ,

**Pull request merged**